odl/trafos/backends/pywt_bindings.py

```diff
@@ -0,0 +1,523 @@
+
+        In 2D:
+
+        ``[aaN, (adN, daN, ddN), ..., (ad1, da1, dd1)]``
+
+        In 3D:
+
+        ``[aaaN, (aadN, adaN, addN, daaN, dadN, ddaN, dddN), ...
+        (aad1, ada1, add1, daa1, dad1, dda1, ddd1)]``
+
+        The abbreviations are "a" for "approximation" and "d" for "detail".
+        The letter position encodes the coefficient type in the respective
+        axis. ``N`` stands for the number of scaling levels.
+    """
+    coeff_arr = np.asarray(coeff)
+    flat_sizes = [np.prod(shp) for shp in shape_list[:-1]]
+    start = 0
+    stop = flat_sizes[0]
+    coeff_list = [coeff_arr[start:stop].reshape(shape_list[0])]
+    ndim = len(shape_list[0])
+    dcoeffs_per_scale = 2 ** ndim - 1
+
+    for fsize, shape in zip(flat_sizes[1:], shape_list[1:]):
+        start, stop = stop, stop + dcoeffs_per_scale * fsize
+        if dcoeffs_per_scale == 1:
+            detail_coeffs = coeff_arr[start:stop]
+        else:
+            detail_coeffs = tuple(c.reshape(shape) for c in
+                                  np.split(coeff_arr[start:stop],
+                                           dcoeffs_per_scale))
+        coeff_list.append(detail_coeffs)
+
+    return coeff_list
+
+
+def pywt_wavelet_decomp(arr, wbasis, pad_mode, nscales, shape_list):
+    """Return the discrete wavelet transform of ``arr``.
+
+    Parameters
+    ----------
+    arr : `array-like`
+        Input array to the wavelet decomposition.
+    wbasis :  string or `pywt.Wavelet`
+        Specification of the wavelet to be used in the transform.
+        If a string is given, it is converted to a `pywt.Wavelet`.
+    pad_mode: string
+        ODL style boundary condition.
+    nscales : int
+        Number of scaling levels to be used in the transform.
+        The maximum number of usable scales can be determined
+        by `pywt.dwt_max_level`.
+    shape_list : `sequence`
+        Sequence containing the shapes of the wavelet (approximation
+        and detail) coefficients at different scaling levels.
+
+    Returns
+    -------
+    out : `numpy.ndarray`
+        Flattened and concatenated coefficient array.
+        The length of the array depends on the shape of the input image to
+        be transformed and on the chosen wavelet basis.
+    """
+    arr = np.asarray(arr)
+    wbasis = pywt_wbasis(wbasis)
+    pywt_mode = PAD_MODES_ODL2PYWT[pad_mode]
+    if arr.ndim == 1:
+        coeff_list = pywt.wavedec(arr, wbasis, pywt_mode, nscales)
+    elif arr.ndim == 2:
+        coeff_list = pywt.wavedec2(arr, wbasis, pywt_mode, nscales)
+    elif arr.ndim == 3:
+        coeff_list = _wavedec3(arr, wbasis, pywt_mode, nscales)
+    else:
+        raise NotImplementedError('no transform available for {} dimensions'
+                                  ''.format(arr.ndim))
+    return pywt_coeff_to_array(coeff_list, shape_list)
+
+
+def pywt_wavelet_recon(coeff, wbasis, pad_mode, shape_list):
+    """Return the discrete wavelet reconstruction from ``coeff``.
```

Comment:
That's new to me. My understanding was that we don't use it for `@property` since they look exactly like attributes, and attributes *are* something, in contrast to methods or functions, which *do* something (return, modify, check, ...).